rfc/JENKINS-41334-parallel-stage-execution.md

```diff
@@ -0,0 +1,76 @@
+## RFC: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334) - Parallel stage execution
+
+* Author: Andrew Bayer
+* Created Date: February 27, 2017
+* Approved Date: n/a
+* Target Release: 1.2
+* JIRA: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334)
+* Status: Draft
+
+### Problem Statement
+
+The `parallel` step does not fit smoothly into Declarative. It doesn't
+allow for parallel stage execution, it doesn't utilize the `agent`
+definition, etc. There needs to be the ability to run `stage`s in
+parallel, with the full power of the Declarative syntax enabled -
+i.e., `agent`, `post`, `when`, etc.
+
+### Syntax
+
+#### Proposed example
+
+```groovy
+stage('foo') {
+    parallel {
+        stage('first') {
+            steps {
+                echo "First branch"
+            }
+        }
+        stage('second') {
+            steps {
+                echo "Second branch"
+            }
+        }
+    }
+}
+```
+
+#### Details
+
+* A `stage` has to contain one and only one of `steps` or
+    `parallel`. They cannot be combined in the same `stage`.
+* `stage` configuration can be done on both a `stage` containing
+    `parallel`, or `stage`s within `parallel`, i.e., `agent`, `when`,
+    `post`, etc.
+* `agent`, `environment`, and `tools` specified on the "parent"
+    `stage` will apply to child `stage`s in the same way that top-level
+    configuration applies to `stage`s currently.
+* Arbitrarily deep nesting of parallel `stage`s would not be
+    allowed. Only `stage`s that themselves are not within a `parallel`
+    would be able to contain further `parallel` `stage`s.
+* `stage`s within a `parallel` would not allow use of the `parallel`
+    step either, so as to prevent visualization confusion.
+* Naming of `parallel` is up for debate.
+
+### Runtime Implementation
+
+Based on input from the Blue Ocean team regarding the two possible runtime
+implementations (actually nesting stages within the `parallel` branches vs 
+simply utilizing the `parallel` branches), we have decided to just use the
+`parallel` branches themselves. This should hopefully require little to no
+changes in Blue Ocean visualization or Bismuth in order to work, meaning only
+changes to the editor will be needed outside of Declarative itself.
+
+Each `parallel` branch will be given the `stage` name specified, with `agent`,
```

Comment:
👎  on this.  If you want to declare a branch, call it a branch, not a stage.  Pre/post can be done, but I do not think we could or should effectively support using metadata with branches, and especially conditional execution at this point.  This is a hack and a potentially dangerous one.   There's a right way to do this, and that's to start looking at a richer execution & status model for pipeline -- but that needs to be provided from a lower level API and then bubble up to the Declarative which takes advantage.